nucypher/policy/maps.py

```diff
@@ -35,12 +35,63 @@
 from nucypher.crypto.kits import UmbralMessageKit
 from nucypher.crypto.powers import DecryptingPower, SigningPower
 from nucypher.crypto.signing import SignatureStamp
-from nucypher.crypto.splitters import signature_splitter
-from nucypher.crypto.umbral_adapter import KeyFrag, PublicKey, Signature
+from nucypher.crypto.splitters import signature_splitter, hrac_splitter, kfrag_splitter
+from nucypher.crypto.umbral_adapter import KeyFrag, VerifiedKeyFrag, PublicKey, Signature
 from nucypher.crypto.utils import keccak_digest, encrypt_and_sign, verify_eip_191
 from nucypher.network.middleware import RestMiddleware
 
 
+class AuthorizedKeyFrag:
+
+    _splitter = BytestringSplitter(
+        hrac_splitter, # HRAC
+        BytestringSplitter((bytes, WRIT_CHECKSUM_SIZE)), # kfrag checksum
+        signature_splitter, # alice signature
+        kfrag_splitter,
+        )
+
+    @staticmethod
+    def _kfrag_checksum(kfrag: Union[KeyFrag, VerifiedKeyFrag]) -> bytes:
+        # TODO: assuming here that KeyFrag and VerifiedKeyFrag have the same byte representation.
+        # It's a little hacky... Should we have `verified_kfrag.unverify()` or something?
+        return keccak_digest(bytes(kfrag))[:WRIT_CHECKSUM_SIZE]
+
+    @classmethod
+    def create(cls, hrac: bytes, kfrag: VerifiedKeyFrag, alice_stamp: SignatureStamp) -> 'AuthorizedKeyFrag':
+
+        # Alice makes plain to Ursula that, upon decrypting this message,
+        # this particular KFrag is authorized for use in the policy identified by this HRAC.
+        kfrag_checksum = cls._kfrag_checksum(kfrag)
+        writ = hrac + kfrag_checksum
+        writ_signature = alice_stamp(writ)
+
+        # The writ and the KFrag together represent a complete kfrag kit: the entirety of
+        # the material needed for Ursula to assuredly service this policy.
+        return cls(hrac, kfrag_checksum, writ_signature, kfrag)
+
+    def __init__(self, hrac: bytes, kfrag_checksum: bytes, writ_signature: Signature, kfrag: VerifiedKeyFrag):
+        self.hrac = hrac
+        self.kfrag_checksum = kfrag_checksum
+        self.writ = hrac + kfrag_checksum
+        self.writ_signature = writ_signature
+        self.kfrag = kfrag
```

Comment:
Should we name this `verified_kfrag`?

It gets a bit confusing between Kfrag, VerifiedKFrag, AuthorizedKeyfrag...

Reply:
Actually at this point it can either be `KeyFrag` or `VerifiedKeyFrag`, depending on whether it `AuthorizedKeyFrag` was created by `from_bytes()` or `create()`. I guess we just need an un-verification in the latter. Not sure if a separate method is worth it, `KeyFrag.from_bytes(bytes(verified_kfrag))` will do the trick.

Reply:
🤯

Reply:
I'm a little confused. Is the type hint of `VerifiedKeyFrag` wrong then...?

Is the 'authorized' dictated by the `writ`? and then it's up to whoever is using this AuthorizedKeyFrag to check whether the key frag included is verified or not? Missing a bit of the larger picture.

Reply:
Yes, as it is called now it should be `Union[KeyFrag, VerifiedKeyFrag]`. But if it actually is a VKF, it will cause problems later to whoever tries to verify it (if for some reason you created an AKF and try to verify without it passing through a serialization-deserialization), so it's better to just un-verify in `create()`.

Reply:
Ok, so `AuthorizedKeyFrag` should only ever take a `KeyFrag` in `__init__` and never a `VerifiedKeyFrag` then. And to achieve that we should add `KeyFrag.from_bytes(bytes(verified_kfrag))` in the `create()` function.

Reply:
Added an un-verifiation step.

Reply:
👍